### ISSUE TITLE
MissingFieldException when calling proceed on 'select-enroll-profile'

### ISSUE THREAD

**Issue by ckasek** [okta/okta-idx-android#196]
### Describe the bug?

Things had been working fine in our app for months and then all of a sudden, I started getting this exception when calling proceed on 'select-enroll-profile':

`kotlinx.serialization.MissingFieldException: Field 'age' is required for type with serial name 'com.okta.idx.kotlin.dto.v1.Authenticator.Settings', but it was missing at path: $.currentAuthenticator.value.settings`

Using the same client ID, calling proceed on 'select-enroll-profile' works fine in our iOS app.  I've tried with both our application as well as the dynamic-app application and see the error with both.

Using network inspector, I see this response body, where the $currentAuthenticator does not have an "Age" object:

```
{
  "version": "1.0.0",
  "stateHandle": "<removed>",
  "expiresAt": "2024-03-12T16:57:38.000Z",
  "intent": "LOGIN",
  "remediation": {
    "type": "array",
    "value": [
      {
        "rel": [
          "create-form"
        ],
        "name": "enroll-profile",
        "href": "https://**********.oktapreview.com/idp/idx/enroll/new",
        "method": "POST",
        "produces": "application/ion+json; okta-version=1.0.0",
        "value": [
          {
            "name": "userProfile",
            "form": {
              "value": [
                {
                  "name": "firstName",
                  "type": "string",
                  "label": "First name",
                  "required": true,
                  "minLength": 1,
                  "maxLength": 50
                },
                {
                  "name": "lastName",
                  "type": "string",
                  "label": "Last name",
                  "required": true,
                  "minLength": 1,
                  "maxLength": 50
                },
                {
                  "name": "email",
                  "type": "string",
                  "label": "Primary email",
                  "required": true,
                  "maxLength": 100
                },
                {
                  "name": "login",
                  "type": "string",
                  "label": "Username",
                  "required": true,
                  "minLength": 5,
                  "maxLength": 100
                }
              ]
            }
          },
          {
            "name": "credentials",
            "type": "object",
            "form": {
              "value": [
                {
                  "name": "passcode",
                  "label": "Enter Password",
                  "required": true,
                  "secret": true
                }
              ]
            },
            "required": true,
            "relatesTo": "$.currentAuthenticator"
          },
          {
            "name": "stateHandle",
            "required": true,
            "value": "<removed>",
            "visible": false,
            "mutable": false
          }
        ],
        "accepts": "application/json; okta-version=1.0.0"
      },
      {
        "rel": [
          "create-form"
        ],
        "name": "select-identify",
        "href": "https://********.oktapreview.com/idp/idx/identify/select",
        "method": "POST",
        "produces": "application/ion+json; okta-version=1.0.0",
        "value": [
          {
            "name": "stateHandle",
            "required": true,
            "value": "<removed>",
            "visible": false,
            "mutable": false
          }
        ],
        "accepts": "application/json; okta-version=1.0.0"
      }
    ]
  },
  "uiDisplay": {
    "type": "object",
    "value": {
      "label": "Sign in",
      "buttonLabel": "Submit"
    }
  },
  "currentAuthenticator": {
    "type": "object",
    "value": {
      "type": "password",
      "key": "okta_password",
      "id": "aut57g4figPC2UffB1d7",
      "displayName": "Password",
      "methods": [
        {
          "type": "password"
        }
      ],
      "settings": {
        "complexity": {
          "minLength": 8,
          "minLowerCase": 1,
          "minUpperCase": 1,
          "minNumber": 1,
          "minSymbol": 1,
          "excludeUsername": true,
          "excludeAttributes": []
        }
      }
    }
  }
  ```

### What is expected to happen?

If "Age" is not guaranteed to be in the response coming from the Okta servers, then the JSON parsing should not fail due to it not being there.

If "Age" is expected to be in the response, then it would seem there is an error on the backend somewhere that results in it not coming back in the response.

### What is the actual behavior?

JSON Parsing fails due to the "Age" object being missing in the Authenticator.Settings object.

### Reproduction Steps?

Using the dynamic app, click the login button, then the sign up button.

### Additional Information?

larger stack trace:

```
2024-03-12 07:57:37.939  5746-5746  DynamicAut...el$proceed com.okta.idx.android                 E  kotlinx.serialization.MissingFieldException: Field 'age' is required for type with serial name 'com.okta.idx.kotlin.dto.v1.Authenticator.Settings', but it was missing at path: $.currentAuthenticator.value.settings
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:93)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$$serializer.deserialize(Responses.kt:115)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$$serializer.deserialize(Responses.kt:115)
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:165)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.IonObject$$serializer.deserialize(Responses.kt:83)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.IonObject$$serializer.deserialize(Responses.kt:83)
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Response$$serializer.deserialize(Responses.kt:66)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Response$$serializer.deserialize(Responses.kt:66)
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
                                                                                                    	at kotlinx.serialization.json.internal.JsonStreamsKt.decodeByReader(JsonStreams.kt:39)
                                                                                                    	at kotlinx.serialization.json.okio.OkioStreamsKt.decodeFromBufferedSource(OkioStreams.kt:65)
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$performRequest$3.invoke(NetworkUtils.kt:64)
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$performRequest$3.invoke(NetworkUtils.kt:63)
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$3.invoke(NetworkUtils.kt:97)
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$3.invoke(NetworkUtils.kt:96)
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$5.invokeSuspend(NetworkUtils.kt:114)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: kotlinx.serialization.MissingFieldException: Field 'age' is required for type with serial name 'com.okta.idx.kotlin.dto.v1.Authenticator.Settings', but it was missing
                                                                                                    	at kotlinx.serialization.internal.PluginExceptionsKt.throwMissingFieldException(PluginExceptions.kt:20)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$Settings.<init>(Responses.kt:130)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$Settings$$serializer.deserialize(Responses.kt:130)
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$Settings$$serializer.deserialize(Responses.kt:130)
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70)
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43) 
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$$serializer.deserialize(Responses.kt:115) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Authenticator$$serializer.deserialize(Responses.kt:115) 
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70) 
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43) 
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70) 
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:165) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.IonObject$$serializer.deserialize(Responses.kt:83) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.IonObject$$serializer.deserialize(Responses.kt:83) 
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70) 
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43) 
                                                                                                    	at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Response$$serializer.deserialize(Responses.kt:66) 
                                                                                                    	at com.okta.idx.kotlin.dto.v1.Response$$serializer.deserialize(Responses.kt:66) 
                                                                                                    	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:70) 
                                                                                                    	at kotlinx.serialization.json.internal.JsonStreamsKt.decodeByReader(JsonStreams.kt:39) 
                                                                                                    	at kotlinx.serialization.json.okio.OkioStreamsKt.decodeFromBufferedSource(OkioStreams.kt:65) 
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$performRequest$3.invoke(NetworkUtils.kt:64) 
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$performRequest$3.invoke(NetworkUtils.kt:63) 
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$3.invoke(NetworkUtils.kt:97) 
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$3.invoke(NetworkUtils.kt:96) 
                                                                                                    	at com.okta.authfoundation.client.internal.NetworkUtilsKt$internalPerformRequest$5.invokeSuspend(NetworkUtils.kt:114) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
```

### SDK Version

I tried both 2.3.0 and 2.4.0 SDK versions with our app.

I used the latest code from master for testing the dynamic-app

**Comment by ckasek** [comment]
The exact place the problem originates from seems to be here - https://github.com/okta/okta-idx-android/blob/master/okta-idx-kotlin/src/main/java/com/okta/idx/kotlin/dto/v1/Responses.kt#L131

settings is optional and defaults to null, but if settings is present, then age and complexity are both required.  As you can see in the response I included above, I am receiving a settings value that contains complexity rules, but no age rules.  This seems to make sense since it is a new user registering, that there would be no age rules since it is a new user with no password history.

**Comment by nishant-mehta** [comment]
We are facing the same issue.

**Comment by nishant-mehta** [comment]
It seems to be related:

https://help.okta.com/oie/en-us/content/topics/releasenotes/production-oie.htm

Password age validation incorrectly appeared on the new user registration window. (OKTA-673824)

Was the code tested?

**Comment by emanor-okta** [comment]
OKTA-710908

@ckasek - Depending on how your application is setup and if it relies on `password` being returned on the initial `/enroll` call, you can uncheck 'Show password field on the first page of the enrollment form' in the Profile enrollment policy and this will allow setting the password when other authenticators are enrolled without this error happening.

**Comment by rajdeepnanua-okta** [comment]
The fix for this issue should be released shortly in okta-idx-kotlin 2.4.1
